### PR TITLE
Correctly handle false on Add/Take from underlying collection in BlockingCollection

### LINKED ISSUE #69816: The BlockingCollection<T> backed by a custom collection doesn't work properly
### Description

I am trying to create a [`BlockingCollection<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1) that rejects duplicate items. For this reason I wrote a custom collection that implements the [`IProducerConsumerCollection<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.iproducerconsumercollection-1) interface, and then passed an instance of this collection to the constructor of the `BlockingCollection<T>`. 

The problem is that the bounding functionality of the `BlockingCollection<T>` seems to brake when a duplicate item has been rejected by the underlying collection. Although the item has been rejected, it still counts towards the [`BoundedCapacity`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.boundedcapacity) limit. The result is that fewer than `BoundedCapacity` elements can be added in the collection, and eventually (after many rejected duplicates) the effective capacity is reduced to just 1.

### Reproduction Steps

```C#
using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;

public class Program
{
    public static void Main()
    {
        BlockingCollection<int> queue = new(new ConcurrentQueueNoDuplicates<int>(), 2);
        AddItem(1); // True
        AddItem(1); // InvalidOperationException
        AddItem(2); // False (unexpected)

        void AddItem(int value)
        {
            try
            {
                Console.WriteLine($"queue.TryAdd({value}): {queue.TryAdd(value)}");
            }
            catch (InvalidOperationException) when (!queue.IsAddingCompleted)
            {
                Console.WriteLine($"queue.TryAdd({value}): InvalidOperationException");
            }
        }
    }

    private class ConcurrentQueueNoDuplicates<T> : IProducerConsumerCollection<T>
    {
        private readonly Queue<T> _queue = new();
        public bool TryAdd(T item)
        {
            lock (_queue)
            {
                if (_queue.Contains(item)) return false;
                _queue.Enqueue(item); return true;
            }
        }
        public bool TryTake(out T item)
        {
            lock (_queue)
            {
                if (_queue.Count == 0) { item = default; return false; }
                item = _queue.Dequeue(); return true;
            }
        }
        public int Count { get { lock (_queue) return _queue.Count; } }
        public bool IsSynchronized => throw new NotImplementedException();
        public object SyncRoot => throw new NotImplementedException();
        public void CopyTo(T[] array, int index) => throw new NotImplementedException();
        public void CopyTo(Array array, int index) => throw new NotImplementedException();
        public IEnumerator<T> GetEnumerator() => throw new NotImplementedException();
        IEnumerator IEnumerable.GetEnumerator() => throw new NotImplementedException();
        public T[] ToArray() => throw new NotImplementedException();
    }
}
```

Output:

```none
queue.TryAdd(1): True
queue.TryAdd(1): InvalidOperationException
queue.TryAdd(2): False
```

[Live demo](https://dotnetfiddle.net/BrE4h8).

### Expected behavior

The `BoundedCapacity` is configured with the value `2`, so it is expected that the two values 1 and 2 can both be stored in the collection. The attempt to add the value 1 twice should be denied, but the state of the collection should remain unchanged. So the subsequent `queue.TryAdd(2)` should return `true`.

### Actual behavior

The `queue.TryAdd(2)` returns `false`.

### Regression?

The actual behavior on .NET Framework 4.8 is the same.

### Known Workarounds

The underlying collection rejects an item by throwing an exception instead of returning `false`.

### Configuration

- .NET 6.0
- Windows 10
- 64-bit operating system, x64-based processor

### Other information

The [source code](https://github.com/dotnet/runtime/blob/56e821243d92f20a6abf7ca61288221f4d164dd1/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs#L480-L483) includes the following comment:

```C#
//TryAdd is guaranteed to find a place to add the element. Its return value depends
//on the semantics of the underlying store. Some underlying stores will not add an already
//existing item and thus TryAdd returns false indicating that the size of the underlying
//store did not increase.
```

The actual behavior of the `BlockingCollection<int>.TryAdd` is not to return `false`, but to throw an exception. The actual behavior is correctly [documented](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.tryadd) in the online docs. My guess is that the specification changed during the development of the class, and the problem is somehow related to that change.

### PR DESCRIPTION
**Pull request by hrrrrustic**
fix #69816
Have some troubles with running tests manually, waiting CI

### CODE REVIEW

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs [hunk 1]
```diff
@@ -734,7 +735,7 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
                             _freeNodes.Release();
                         }
                     }
-                    else if (removeFaulted)
+                    else if (removeFaulted ||!removeSucceeded && !removeFaulted)
```
Comment (hrrrrustic) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:738]:
```suggestion
                    else if (removeFaulted || !removeSucceeded)
```

Comment (hrrrrustic) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:738]: That's folding into just `true` 🤔

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs [hunk 2]
```diff
@@ -467,47 +467,29 @@ private bool TryAddWithNoTimeValidation(T item, int millisecondsTimeout, Cancell
...
-                    }
-                    if (addingSucceeded)
-                    {
-                        //After adding an element to the underlying storage, signal to the consumers
-                        //waiting on _occupiedNodes that there is a new item added ready to be consumed.
-                        _occupiedNodes.Release();
-                    }
-                    else
-                    {
+                    if(!addingSucceeded)
```
Comment (gfoidl) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:483]: Nit:
```suggestion
                    if (!addingSucceeded)
```

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs [hunk 3]
```diff
@@ -467,47 +467,29 @@ private bool TryAddWithNoTimeValidation(T item, int millisecondsTimeout, Cancell
...
-                // 2- _collection.TryAdd succeeded
-                // 3- _collection.TryAdd returned false
-                // so we put the decrement code in the finally block
+                //TryAdd is guaranteed to find a place to add the element. Its return value depends
+                //on the semantics of the underlying store. Some underlying stores will not add an already
+                //existing item and thus TryAdd returns false indicating that the size of the underlying
+                //store did not increase.
+                bool addingSucceeded = false;
+
                 try
```
Comment (hrrrrustic) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:476]: This now is pretty similar to how `TryTake` looks like

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs [hunk 4]
```diff
@@ -734,7 +714,7 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
                             _freeNodes.Release();
                         }
                     }
-                    else if (removeFaulted)
```
Comment (hrrrrustic) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:737]: New condition should be `removeFaulted || !removeSucceeded && !removeFaulted` => `removeFaulted || !removeSucceeded`, but as we already in else block, `removeSucceeded` is definitely `false` and whole condition folds in single `true`

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs [hunk 5]
```diff
@@ -467,47 +467,29 @@ private bool TryAddWithNoTimeValidation(T item, int millisecondsTimeout, Cancell
...
-                    try
-                    {
-                        //The token may have been canceled before the collection had space available, so we need a check after the wait has completed.
-                        //This fixes bug #702328, case 2 of 2.
-                        cancellationToken.ThrowIfCancellationRequested();
-                        addingSucceeded = _collection.TryAdd(item);
-                    }
-                    catch
-                    {
-                        //TryAdd did not result in increasing the size of the underlying store and hence we need
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:493]: Would it make sense to salvage some of these comments in the new location?

File: src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:483]: Yes please.

**Review by eiriktsarpalis — COMMENTED** [review]
Other than a few minor nits, this looks good to me.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Pull request merged**